packages/app-metadata-controller/src/AppMetadataController.ts

```diff
@@ -0,0 +1,184 @@
+    currentMigrationVersion: 0,
+  });
+
+/**
+ * Returns the state of the {@link AppMetadataController}.
+ */
+export type AppMetadataControllerGetStateAction = ControllerGetStateAction<
+  typeof controllerName,
+  AppMetadataControllerState
+>;
+
+/**
+ * Actions exposed by the {@link AppMetadataController}.
+ */
+export type AppMetadataControllerActions = AppMetadataControllerGetStateAction;
+
+/**
+ * Event emitted when the state of the {@link AppMetadataController} changes.
+ */
+export type AppMetadataControllerStateChangeEvent = ControllerStateChangeEvent<
+  typeof controllerName,
+  AppMetadataControllerState
+>;
+
+export type AppMetadataControllerEvents = AppMetadataControllerStateChangeEvent;
+
+/**
+ * Actions that this controller is allowed to call.
+ */
+type AllowedActions = never;
+
+/**
+ * Events that this controller is allowed to subscribe.
+ */
+type AllowedEvents = never;
+
+/**
+ * Messenger type for the {@link AppMetadataController}.
+ */
+type AppMetadataControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  AppMetadataControllerActions | AllowedActions,
+  AppMetadataControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * {@link AppMetadataController}'s metadata.
+ *
+ * This allows us to choose if fields of the state should be persisted or not
+ * using the `persist` flag; and if they can be sent to Sentry or not, using
+ * the `anonymous` flag.
+ */
+const controllerMetadata = {
+  currentAppVersion: {
+    persist: true,
+    anonymous: true,
+  },
+  previousAppVersion: {
+    persist: true,
+    anonymous: true,
+  },
+  previousMigrationVersion: {
+    persist: true,
+    anonymous: true,
+  },
+  currentMigrationVersion: {
+    persist: true,
+    anonymous: true,
+  },
+};
+
+/**
+ * The AppMetadata controller stores metadata about the current extension instance,
+ * including the currently and previously installed versions, and the most recently
+ * run migration.
+ *
+ */
+export default class AppMetadataController extends BaseController<
+  typeof controllerName,
+  AppMetadataControllerState,
+  AppMetadataControllerMessenger
+> {
+  /**
+   * Constructs a AppMetadata controller.
+   *
+   * @param options - the controller options
+   * @param options.state - Initial controller state.
+   * @param options.messenger - Messenger used to communicate with BaseV2 controller.
+   * @param options.currentMigrationVersion
+   * @param options.currentAppVersion
+   */
+  constructor({
+    state = {},
+    messenger,
+    currentAppVersion = '',
+    currentMigrationVersion = 0,
+  }: AppMetadataControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      state: {
+        ...getDefaultAppMetadataControllerState(),
+        ...state,
+      },
+      messenger,
+    });
+
+    this.#maybeUpdateAppVersion(currentAppVersion);
+
+    this.#maybeUpdateMigrationVersion(currentMigrationVersion);
```

Comment:
In case my comment above makes sense, then these would also be removed since no update needs to happen
```suggestion
```

Reply:
Oh, I see that the extension is already using it this way. You can discard this comment